juju/juju
# new apiserver modelgeneration: add, switch, advance

## Description of change

- New apiserver for modelgeneration, includes AddGeneration, AdvanceGeneration, and SwitchGeneration, and related unit tests.
- Unit tests for api/modelgeneration added for  AddGeneration, AdvanceGeneration, and SwitchGeneration.
- Updated cmd and api for AddGeneration, AdvanceGeneration, and SwitchGeneration to take model tags.
- Moved MockResources and MockAuthorizer to api/base for multiple package use.
- Added MockAPICallerCloser.
- Related changes for CancelGeneration will be forth coming.

## QA steps

1. export JUJU_DEV_FEATURE_FLAGS=generations
2. juju bootstrap localhost
2052.  juju deploy percona-cluster mysql
2056.  juju add-generation
    target generation set to next
2057.  juju switch-generation current
    changes dropped and target generation set to current
2059.  juju advance mysql
    ERROR next generation is not active
2060.  juju switch-generation next
    target generation set to next
2061.  juju advance mysql
2062.  juju switch-generation next
    ERROR cannot switch to next generation, as none exists
2063.  juju add-unit mysql
2064.  juju add-generation
    target generation set to next
2065.  juju advance mysql/1
2068.  juju advance mysql/0
2069.  juju switch-generation next
    ERROR cannot switch to next generation, as none exists
2070.  juju add-generation
    target generation set to next
071.  juju advance mysql/0
2072.  juju advance mysql
2073.  juju switch-generation next
    ERROR cannot switch to next generation, as none exists
4. run `juju-db.bash`
2. db.generations.find(), should report 3 docs with: active = false, completed with an non zero NumberLong.

## Documentation changes

In the future
